dataset/src/main/scala/frameless/TypedDataset.scala

```diff
@@ -325,6 +325,28 @@ class TypedDataset[T] protected[frameless](val dataset: Dataset[T])(implicit val
   def makeUDF[A1: TypedEncoder, A2: TypedEncoder, A3: TypedEncoder, A4: TypedEncoder, A5: TypedEncoder, R: TypedEncoder](f: (A1, A2, A3, A4, A5) => R):
   (TypedColumn[T, A1], TypedColumn[T, A2], TypedColumn[T, A3], TypedColumn[T, A4], TypedColumn[T, A5]) => TypedColumn[T, R] = functions.udf(f)
 
+
+  /** Prepends a new column to the Dataset.
+    * {{{
+    *   case class X(i: Int, j: Int)
+    *   val f: TypedDataset[X] = TypedDataset.create(X(1,1) :: X(1,1) :: X(1,10) :: Nil)
+    *   val fNew: TypedDataset[(Int,Int,Boolean)] = f.withColumn(f('j) === 10)
+    * }}}
+    */
+  def withColumn[A: TypedEncoder, H <: HList, FH <: HList, Out](ca: TypedColumn[T, A])(
+    implicit
+    g: Generic.Aux[T, H],
+    init: Prepend.Aux[H, A :: HNil, FH],
+    t: Tupler.Aux[FH, Out],
```

Comment:
@OlivierBlanvillain, I agree! However, this is a generic behavior we encounter throughout frameless. For example, same happens with select(). In spark select preserves the column names (unless you apply an expression), here, after a select() we fall back to tuples. I think a nice overall solution is to have the default fallback Product to be a labeled generic hlist instead of a Scala tuple. Doing so, after selects and column appends, all the column names can be reserved. Definitely this will required many changes, so I believe it deserves its own PR. For now I think it makes the behavior kind of uniform and expected if we fall back to tuples anytime we remove, append, or project columns.